core/modules/node/node.pages.inc

```diff
@@ -444,7 +444,21 @@ function node_form_submit($form, &$form_state) {
   if ($node->nid) {
     $form_state['values']['nid'] = $node->nid;
     $form_state['nid'] = $node->nid;
-    $form_state['redirect'] = node_access('view', $node) ? 'node/' . $node->nid : '<front>';
+
+    // Determine whether to redirect user to full page display.
+    $form_state['redirect'] = '<front>';
+    if (node_access('view', $node)) {
+      $form_state['redirect'] = 'node/' . $node->nid;
+
+      // If node type set to prevent direct viewing of full page URL
+      // stay on edit page if user is not permitted.
+      $hidden_path = config_get('node.type.' . $node->type, 'settings.hidden_path');
```

Comment:
Instead of loading the config file directly, you should use `node_type_get_type()`, which is cached.